### PR TITLE
add ros parameters file to node context

### LINKED ISSUE #59: Custom parameter parsing not working
The custom logic for parsing parameters in the gazebo_plugin does not work as expect. When launching the system with the following parameters:
```
controller_manager:
  ros__parameters:
    update_rate: 100  # Hz

    joint_trajectory_controller:
      type: joint_trajectory_controller/JointTrajectoryController

    joint_state_controller:
      type: joint_state_controller/JointStateController

joint_trajectory_controller:
  ros__parameters:
    joints:
      - joint1
      - joint2
```

I expect to see the parameters set at the controller manager, however `ros2 param list` yields nothing:
```
 ➭ ros2 param list
...
/controller_manager:
  use_sim_time

```

### PR DESCRIPTION
**Pull request by Karsten1987**
fixes #59 

With this PR, the actual parameters are being set:

```
 ➭ ros2 param list
...
/controller_manager:
  joint_state_controller.type
  joint_trajectory_controller.type
  update_rate
  use_sim_time
```

While implementing this, I found actually a second bug in which all specified controllers in the loaded yaml file would be loaded. It didn't show in the example, however this would necessarily lead to resource conflicts when more than one controller use the same command interface.

### CODE REVIEW

File: gazebo_ros2_control/src/gazebo_ros2_control_plugin.cpp [hunk 1]
```diff
@@ -214,214 +218,70 @@ void GazeboRosControlPlugin::Load(gazebo::physics::ModelPtr parent, sdf::Element
...
+  auto parameter_map = rclcpp::parameter_map_from(params_struct);
+  rcl_yaml_node_struct_fini(params_struct);
+  auto it = parameter_map.find("/controller_manager");
+  if (it != parameter_map.end()) {
+    impl_->controller_manager_->set_parameters(it->second);
+  } else {
+    RCLCPP_INFO(
+      impl_->model_nh_->get_logger(),
+      "No parameters for /controller_manager found, setting defaults");
+  }
```
Comment (ahcorde) [inline gazebo_ros2_control/src/gazebo_ros2_control_plugin.cpp:276]:
```suggestion
  std::vector<std::string> start_controllers = {};
  std::vector<std::string> stop_controllers = {};
  auto parameter_map = rclcpp::parameter_map_from(params_struct);
  rcl_yaml_node_struct_fini(params_struct);
  auto it = parameter_map.find("/controller_manager");
  if (it != parameter_map.end()) {
    for (auto param : it->second) {
      if (param.get_name() == "update_rate") {
        // Decide the plugin control period
        impl_->control_period_ = rclcpp::Duration(
          std::chrono::duration_cast<std::chrono::nanoseconds>(
            std::chrono::duration<double>(
              1.0 / static_cast<double>(param.get_value_message().integer_value))));

        // Check the period against the simulation period
        if (impl_->control_period_ < gazebo_period) {
          RCLCPP_ERROR_STREAM(
            impl_->model_nh_->get_logger(),
            "Desired controller update period (" << impl_->control_period_.seconds() <<
              " s) is faster than the gazebo simulation period (" <<
              gazebo_period.seconds() << " s).");
        } else if (impl_->control_period_ > gazebo_period) {
          RCLCPP_WARN_STREAM(
            impl_->model_nh_->get_logger(),
            " Desired controller update period (" << impl_->control_period_.seconds() <<
              " s) is slower than the gazebo simulation period (" <<
              gazebo_period.seconds() << " s).");
        }
      }
      else {
        std::string controller_name = param.get_name();
        std::string controller_type = param.as_string();
        std::size_t found = controller_name.find(".");
        if (found != std::string::npos) {
          controller_name = controller_name.substr(0, found);
        }

        auto controller = impl_->controller_manager_->load_controller(
          controller_name,
          controller_type);

        auto it_controller = parameter_map.find("/"+controller_name);
        if (it_controller != parameter_map.end()) {
          controller->get_node()->set_parameters(it_controller->second);
        }
        impl_->controllers_.push_back(controller);
        controller->configure();
        start_controllers.push_back(controller_name);
      }
    }
```

File: gazebo_ros2_control/src/gazebo_ros2_control_plugin.cpp [hunk 2]
```diff
@@ -214,214 +218,70 @@ void GazeboRosControlPlugin::Load(gazebo::physics::ModelPtr parent, sdf::Element
...
+
```
Comment (ahcorde) [inline gazebo_ros2_control/src/gazebo_ros2_control_plugin.cpp:277]:
```suggestion
  auto switch_future = std::async(
    std::launch::async,
    &controller_manager::ControllerManager::switch_controller,
    impl_->controller_manager_,
    start_controllers, stop_controllers,
    2, true, rclcpp::Duration(0, 0));  // STRICT_: 2
  while (std::future_status::timeout == switch_future.wait_for(std::chrono::milliseconds(100))) {
    impl_->controller_manager_->update();
  }
  controller_interface::return_type result = switch_future.get();
  if (result != controller_interface::return_type::SUCCESS) {
    RCLCPP_ERROR(impl_->model_nh_->get_logger(), "Error initializing the joint_state_controller");
  }
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
@Karsten1987 I was testing your contribution but I was not able to run the example or get the `/joint_states`. You didn't launch the controllers.

**Comment by Karsten1987** [comment]
> You didn't launch the controllers.

That's on purpose. As mentioned above, not all controllers can be loaded/configured at the same time due to resource conflicts. Also I believe it's not good practice to launch them within the gazebo plugin. I'd rather use either the spawner functionality (once available) or load them explicitly via services through a launch file.

**Comment by ahcorde** [comment]
I will have a look to the PR https://github.com/ros-controls/ros2_control/pull/310

anyhow the conditional to get the `udpate_rate` is required.

**Comment by Karsten1987** [comment]
you don't necessarily need the spawner PR. You should be able to use `ros2 control load_start_controller` in the launch file to start the controllers.

**Comment by Karsten1987** [comment]
> anyhow the conditional to get the udpate_rate is required.

done in 1047c22

**Comment by ahcorde** [comment]
@Karsten1987 

I was testing the `ros2 control` commands. I was able to start the `joint_state_controller`

```bash
ros2 control load_start_controller joint_trajectory_controller` 
``` 

But I'm not able to load the `joint_trajectory_controller`. Digging about what it's happening here. The `/joint_trajectory_controller` parameters are not set.

```bash
ros2 param get /joint_trajectory_controller  joints
String values are: []
```

You should set up the parameters for the controller in the code.

**Comment by Karsten1987** [comment]
There's currently no clear way to set parameters to a gazebo plugin's node. So I had to do this manually by parsing the ros parameter's file and add the arguments to the node's context. That's certainly not a recommended way, but the only feasible at the moment.

/cc @chapulina @jacobperron

**Comment by joshnewans** [comment]
I just wanted to add my 2c and say that as someone fairly new to all this, I found it quite confusing that the plugin loaded my controllers on its own as I expected a flow similar to my real robot where I used `ros2 control load_start_controller xxx`. The new behaviour resulting from this PR seemed more intuitive to me.

This PR resolved an issue I was having where the `controller_manager` and its child controllers did not obey the `use_sim_time` parameter I specified in the yaml file (and as I understand it there is currently no way to globally enable `use_sim_time`).

**Comment by Karsten1987** [comment]
@ahcorde friendly ping.

**Comment by ahcorde** [comment]
I created a PR which builds on top of this PR load the controllers in the launch files: https://github.com/ros-simulation/gazebo_ros2_control/pull/64

**Review by ahcorde — APPROVED** [review]

**Pull request merged**